test/e2e/main_test.go

```diff
@@ -0,0 +1,61 @@
+// +build e2e
+
+/*
+Copyright 2019 The Knative Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"reflect"
+	"runtime"
+	"testing"
+
+	"github.com/knative/eventing/test"
+)
+
+// channelTestMap indicates which test cases we want to run for a given CCP.
+var channelTestMap = map[string][]func(t *testing.T){
+	"in-memory-channel": []func(t *testing.T){
+		TestSingleBinaryEvent,
+		TestSingleStructuredEvent,
+		TestEventTransformation,
+		TestChannelChain,
+		TestDefaultBrokerWithManyTriggers,
+	},
+
+	// "gcp-pubsub": []func(t *testing.T){
+	// 	TestEventTransformation,
+	// 	TestDefaultBrokerWithManyTriggers,
+	// },
+}
+
+func TestMain(t *testing.T) {
```

Comment:
1. Forcing running the test from TestMain makes things more complicated. You're already calling a Setup function in each test, use that to setup whatever you need for the test. This way you can still run any test(s) with a simple go test.

1. As discussed offline, eliminate this hardcoded test mapping. Unless there's a good reason (e.g., super slow tests), we want to run all tests for all CCPs. For tests that are exactly the same (but only some sort of config changes), use a helper function and call that function from an explicit test function.

1. Consider splitting the e2e tests into different packages, one for each provisioner. This way you can still use go test without flags to run any combination of tests you'd like.

Reply:
1. It's actually not forcing running the test from TestMain. Users can still run all the tests or a single test with `go test`, the way to run them is exactly the same as before, more information can be found in the updated `README` file. TestMain is just an entry for running all predefined test cases.

2. There are another two reasons for me to make the test mapping decision:
i. Some test cases we have now are is irrelevant with provisioner. 
For example, `TestDefaultBrokerWithManyTriggers` is using the default `Broker` that is automatically created with the `DefaultClusterChannelProvisioner` and is not using a provisioner. Running this case will 3 provisioners will simply repeat it 3 times. 
We can definitely change the `DefaultClusterChannelProvisioner` with a ConfigMap, but it should be in another test case.
ii. Some of the channels might share the same feature but not all.
For example, both Kafka and GCP PubSub channel support persistence but not in-memory or another new channel in the future. If we simply separate test cases for channels through different packages, we will have two same test cases. And if we put it under the common package, we still need a way to identify which channels it applies to, which I think is better to manage at a common place rather than its individual setup.

3. We should definitely split the e2e tests into different packages, but probably not one for each provisioner due to the fact in 2-ii.

I can set up a meeting for discussion if necessary.

Reply:
Another thing is, if we don't have an entry to schedule the tests, it will be hard to test against multiple provisioners in parallel through one single `go test` command. 
For example, we can consider a command like this - `go test -v -tags=e2e -count=1 ./test/e2e -run ^TestSingleBinaryEvent$ -clusterChannelProvisioners in-memory-channel,gcp-pubsub`, in which we want to run the test case with two provisioners. We can definitely write a `for loop`, but then `t.Parallel()` will only apply to this test case itself, not the different CCPs.

If we decide not to support multiple provisioners through one single `go test` command, for every new channel, we'll have to add a `go_test_e2e -timeout=20m ./test/e2e ....` command, which makes our tests 20 minutes longer.

Reply:
For now, the test mapping is hardcoded in the `main_test.go` file. In the future, we can move it to a separate file to make it more configurable.

Reply:
> 1. It's actually not forcing running the test from `TestMain`. Users can still run all the tests or a single test with `go test`, the way to run them is exactly the same as before, more information can be found in the updated `README` file. TestMain is just an entry for running all test cases predefined for the scenario they apply to.

Please update the README to reflect that. Right now it only mentions `TestMain` as the only way to run the tests.

> 2. Aside from the over-testing issue, there are another two reasons for me to make the test mapping decision:
> 
> * Some test cases we have now are irrelevant with provisioner.
>   For example, `TestDefaultBrokerWithManyTriggers` is using the default `Broker` that is automatically created by Eventing. It's using the `DefaultClusterChannelProvisioner` but not a provisioner provided by user. Running this case will 3 provisioners will simply repeat it 3 times.
>   We can definitely change the `DefaultClusterChannelProvisioner` with a ConfigMap, but it should be in another test case.

That should be solved by explicitly creating a test function and using a helper.

> * Some of the channels might share the same feature but not all.
>   For example, both Kafka and GCP PubSub channel support persistence but not in-memory channel or another new channel in the future. If we want to write test for this feature, and simply separate test cases for channels through different packages, we will have two same test cases. And if we put it under the common package, we still need a way to identify which channels it applies to, which I think is better to manage at a common place rather than its individual setup.

Sorry, this is not clear to me. Let's sync offline.

Reply:
Synced offline.